rust/src/database/verkle/crypto/opening.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use banderwagon::Fr;
+use ipa_multipoint::{
+    lagrange_basis::LagrangeBasis,
+    multiproof::{MultiPoint, ProverQuery, VerifierQuery},
+    transcript::Transcript,
+};
+use verkle_trie::constants::{CRS, PRECOMPUTED_WEIGHTS};
+
+use crate::database::verkle::crypto::{Commitment, Scalar};
+
+/// A proof demonstrating that a [`Commitment`] contains certain values at specific positions.
+///
+/// It uses the Inner Product Argument (IPA) proof system for Pedersen commitments.
+/// Details: https://dankradfeist.de/ethereum/2021/07/27/inner-product-arguments.html
+pub struct Opening {
+    pub(crate) proof: ipa_multipoint::multiproof::MultiPointProof,
+}
+
+#[cfg_attr(not(test), expect(unused))]
+impl Opening {
+    /// Creates a new opening for a given commitment, proving that it contains the
+    /// value at the specified position. The opening can then be verified using
+    /// [`Self::verify_single`].
+    ///
+    /// For the opening to be valid, the `values` slice must contain the same sequence
+    /// of scalars that were used to create the commitment.
+    ///
+    /// NOTE: This is fairly expensive operation (tens of milliseconds).
+    /// Prefer to open multiple values or even multiple commitments at once, if possible.
+    pub fn open_single(commitment: &Commitment, values: &[Scalar], position: u8) -> Self {
```

Comment:
nit: `Opening::open_single` sounds like you open the opening, but you are actually creating it. Maybe `for_single` would be more fitting. Same for `open_all`.

Reply:
Yes I thought `new_all` sounds weird, but `for_all` is not too bad - renamed!